elements/nuxeo-document-bulk-actions/nuxeo-edit-documents-button.js

```diff
@@ -0,0 +1,613 @@
+   * Flattens the provided `data` structure (that can have multiple levels) into a single map-like structure, mapping
+   * bound paths to values. Accepts a `prefix` string to be added on all keys of the map. A `flattenedProperties`
+   * parameter can also be passed so that the entries are added to it incrementally.
+   */
+  _flattenProperties(data, prefix, flattenedProperties = {}) {
+    // XXX Avoid flattening batch upload blob values (for blob properties)
+    if (data['upload-batch'] && data['upload-fileId']) {
+      flattenedProperties[prefix] = data;
+      return flattenedProperties;
+    }
+    Object.keys(data).forEach((key) => {
+      const value = data[key];
+      const propertyPath = prefix ? `${prefix}.${key}` : key;
+      if (value instanceof Object && !Array.isArray(value)) {
+        this._flattenProperties(value, propertyPath, flattenedProperties);
+      } else {
+        flattenedProperties[propertyPath] = value;
+      }
+    });
+    return flattenedProperties;
+  }
+
+  /**
+   * Validates and submits the form changes to the server.
+   */
+  _save() {
+    this._setSaving(true);
+    if (!this._validate()) {
+      // validation failed, should not proceed
+      this._setSaving(false);
+      return;
+    }
+    const bulkLayout = this.$$('nuxeo-layout').element;
+    const flattenedProperties = this._flattenProperties(bulkLayout.document.properties, 'document.properties');
+    let properties = '';
+    // go through each of the property paths that exist in the document of the layout
+    Object.keys(flattenedProperties).forEach((boundPath) => {
+      // get the element bound to the property
+      const boundElement = this._getBoundElementFromPath(boundPath);
+      // get the bulk widget wrapper element
+      const bulkWidget = this._getBulkWidget(boundElement);
+      // get the path without the `document.properties.` part, replacing the `.` for `/` (complex fields)
+      const path = boundPath
+        .replace(/^(document\.properties\.)/, '')
+        .split('.')
+        .join('/');
+      if (bulkWidget.updateMode === 'replace') {
+        let value = bulkLayout.get(boundPath);
+        if (this._shouldStringifyValue(value)) {
+          value = JSON.stringify(value);
+        }
+        properties = `${properties}${path}=${value}\n`;
+      } else if (bulkWidget.updateMode === 'remove') {
+        properties = `${properties}${path}=\n`;
+      }
+    });
+    this.input = this.documents;
+    this.params = { properties };
+    super._execute().finally(() => this.fire('refresh'));
+    this.$.dialog.toggle();
+    this._setSaving(false);
+  }
+
+  /**
+   * Returns a boolean value according to whether the provided property `value` should be stringified or not.
+   */
+  _shouldStringifyValue(value) {
+    // check if value is an array
+    if (Array.isArray(value)) {
+      // should stringify the array if all the entries in it are objects
+      return value.every((entry) => entry instanceof Object);
```

Comment:
If we have an array with mixed objects and primitive types I guess we won't know what to do, right? Perhaps this is not even possible or expected? If so, using `some` instead would be more efficient I think:
```suggestion
      return value.some((entry) => entry instanceof Object);
```

Reply:
I don't think there's a scenario where an `Array` can have entries of more than one data type, so yes, it really doesn't make sense to check every single entry.